### PR TITLE
🛑  DNM Deserialization: zero-copy merge subframes when possible

### LINKED ISSUE #5107: Unnecessary deep copy causes memory flare on network comms
distributed git tip, Linux x64

```python
import distributed
import numpy

c = distributed.Client(n_workers=4, threads_per_worker=1, memory_limit="2.8 GiB")
f = c.submit(numpy.random.random, 2 ** 27)  # 1 GiB
c.replicate(f, 4)
# Alternative to replicate, identical effect
# futures = [c.submit(lambda x: None, f, pure=False, workers=[w]) for w in c.has_what()]
```
## Expected behaviour
Thanks to pickle 5 buffers, the peak RAM usage on each worker is 1 GiB

## Actual behaviour
I can see on the dashboard the RAM of all workers that receive the computed future over the network briefly flare up to 2 GiB and then settle down at 1 GiB.
On stderr I read:
```
distributed.worker - WARNING - Unmanaged memory use is high. This may indicate a memory leak or the memory may not be released to the OS; see https://distributed.dask.org/en/latest/worker.html#memtrim for more information. -- Unmanaged memory: 2.07 GiB -- Worker memory limit: 2.80 GiB
distributed.worker - WARNING - Unmanaged memory use is high. This may indicate a memory leak or the memory may not be released to the OS; see https://distributed.dask.org/en/latest/worker.html#memtrim for more information. -- Unmanaged memory: 2.07 GiB -- Worker memory limit: 2.80 GiB
distributed.worker - WARNING - Unmanaged memory use is high. This may indicate a memory leak or the memory may not be released to the OS; see https://distributed.dask.org/en/latest/worker.html#memtrim for more information. -- Unmanaged memory: 2.07 GiB -- Worker memory limit: 2.80 GiB
```
If I reduce the memory_limit to 2 GiB, the workers get killed off.

The sender worker is unaffected by the flaring.

I tested on Python 3.8 and 3.9 and on protocols tcp://, ws:// and ucx:// and all are equally affected.

### PR DESCRIPTION
**Pull request by gjoseph92**
When merging sub-frames before deserializing, if all the sub-frames are memoryviews backed by the same underlying buffer, "merge" them without copying by making a new memoryview pointing to the same buffer, instead of always copying.

Running @crusaderky's reproducer from https://github.com/dask/distributed/issues/5107 on `main`:
before
With this PR:
after

Thinking more about my comment https://github.com/dask/distributed/issues/5107#issuecomment-885227654 (tldr: is copying unavoidable because we have to convert non-contiguous memory into contiguous memory?), I realized that in many cases, all the frames already _are_ in contiguous memory thanks to https://github.com/dask/distributed/pull/4506, and they're also zero-copy memoryviews thanks to https://github.com/dask/distributed/pull/3980 (go @jakirkham!). So merging frames in that case could become a bookkeeping operation, where we just make a view of the buffer spanning all the sub-frames we want to merge.

This appears to work but also feels hacky/brittle to me. It requires calculating and passing down offsets into the underlying buffer, and kinda makes some assumptions about how the chain of calling functions works. I might prefer:
* Reorganizing the functions such that [`unpack_frames`](https://github.com/gjoseph92/distributed/blob/1869b1899f690b7556db0a463bc639322ede548e/distributed/protocol/utils.py#L58-L83) also handles merging frames, since we already have the original buffer and all of the offsets we need there. However, that would require deserializing headers within `unpack_frames`, which is probably just as much of a mess.
* A wrapper class for `memoryview` that tracks its index when sliced, and logic to merge those. I think we'd have to write this in C/Cython though, otherwise it wouldn't expose the buffer protocol.
* Some way to get the index/slice a `memoryview` object is pointing to within its buffer. But I believe there's no Python API for this.

cc @madsbk

- [x] Closes #5107
- [x] Tests added / passed
- [x] Passes `black distributed` / `flake8 distributed` / `isort distributed`

### CODE REVIEW

**Comment by jakirkham** [comment]
Yeah I think reworking serialization to be a bit more efficient is one of the things that Mads was looking into. This may involve moving to [msgspec]( https://github.com/jcrist/msgspec ). Though should note that Mads is on PTO atm (so not trying to bug him 😉)

**Comment by gjoseph92** [comment]
Using msgspec would make things easier, especially since it'll give us a `memoryview` when decoding an extension type, which will make this zero-copy logic straightforward. I'm looking forward to that.

Since Mads is going to be refactoring all this anyway, I'd advocate for getting this in as is then, though it's a bit arcane. I have a feeling it'll help memory performance quite a lot. I think the main reason why https://github.com/dask/distributed/pull/4967 helped with memory pressure so much is that transfers were creating these memory spikes, and it just cut down the number of transfers. Making transfers actually zero-copy (at least for NumPy arrays; not sure about pandas yet?) may make high-communication workloads run much more smoothly.

File: distributed/protocol/core.py
Comment (crusaderky) [inline distributed/protocol/core.py:111]: 5% more readable
```suggestion
                            if not isinstance(f, memoryview) or f.obj is not obj:
```

File: distributed/protocol/serialize.py
Comment (crusaderky) [inline distributed/protocol/serialize.py:462]:
```suggestion
    where the ``frames`` start (in bytes, not frame counts).
```

File: distributed/protocol/serialize.py
Comment (crusaderky) [inline distributed/protocol/serialize.py:493]: You can remove the quote marks from the type annotations if your module starts with ``from __future__ import annotations`` (note: won't work with scheduler.py due to cythonization)

**Review by crusaderky — APPROVED** [review]

File: distributed/protocol/serialize.py
Comment (gjoseph92) [reply distributed/protocol/serialize.py]: Good point, I was just lazy :)

**Comment by gjoseph92** [comment]
Hey @crusaderky any idea why these CI logs are truncated? I'm seeing most of the `not ci1` tests fail in `test_serialize_bytes`, but can't see the actual error. See for example https://github.com/dask/distributed/runs/3195109068.

**Comment by crusaderky** [comment]
@gjoseph92 looks like the whole docker container fell over halfway through. Either that or some issue in the github wrapper around it.

**Comment by jrbourbeau** [comment]
GitHub actions was down for a bit yesterday. @gjoseph92 could you push an empty commit to rerun CI?

**Comment by crusaderky** [comment]
Maybe it's a segmentation fault that is triggered by repr() which in turn is triggered by the printing on the stack trace? I would try to deliberately cause one to see if it spells out SIGSEGV explicitly or if you get an equally nebulous output.

**Comment by mrocklin** [comment]
Gabe how confident are you that this PR won't have unintended
consequences?  My gut reaction would be to wait on this one just because
it's so core and doesn't solve a current critical issue.

**Comment by jakirkham** [comment]
If the question is when should we merge this, would it make sense to do after cutting the release?

Should add Mads is getting back next week. So he may have thoughts on how to build on this in cutting down on memory usage.

**Comment by gjoseph92** [comment]
Yes certainly don't merge this until after the release. Definitely seems like there's a segfault here, which annoyingly I haven't reproduced locally yet on macOS.

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by gjoseph92** [comment]
This should get another review. The failing tests were because the whole buffer underlying the frames may have a short prefix from `pack_frames` (so `frames[0]` doesn't start at byte 0 in the buffer), which we need to pass all the way down.

This made me think of a strange (I think impossible actually) edge case: when `frames` contains memoryviews with different underlying buffers, _and_ that initial prefix-offset is passed in. If they're different buffers, it's unlikely they both need the same prefix offset. I say this is impossible because real comms don't mix-and-match like this AFAIK. Currently:
* TCP: all frames are memoryviews of the same buffer (so same prefix)
* websockets: all frames are separate bytestrings (not memoryviews so irrelevant)
* UCX: zero-copy slices of a NumPy array or bytearray? Not totally sure; seems consistent though (not memoryviews so irrelevant)

Nonetheless I didn't like leaving this case around since it could fail in very weird ways, so I added logic to copy these subframes that don't match up just in case: ab6119a1bfe5360a0915b0f19f83a5dcd586eff9. But I've changed my mind again and decided to just make this case an AssertionError, because:
1. as explained above, it shouldn't happen in real use
2. if some future change or Comm implementation makes it happen, that should be an explicit failure instead of a silent performance degradation.

I'll be out for the next week, so if consensus is to do the copy instead of the assert, just revert e2da6101770956afa18f2e604e68627a3e3771c1.

This has become more complicated than I wanted given that we're hopefully redoing all this soon. I really wish there was some way to get the current index into the underlying buffer that a memoryview was pointing to. That would eliminate all this passing down offsets and be more reliable. Is there any way I'm missing to do this without writing a C extension to do pointer arithmetic on `((PyMemoryViewObject *)x)->view.buf` vs `((PyMemoryViewObject *)memoryview(x.obj))->view.buf`?

**Comment by jakirkham** [comment]
UCX uses a mix of NumPy arrays and RMM `DeviceBuffer`s depending on if it is in CPU or GPU memory. That said, wouldn't worry about that. We can probably refresh that layer to benefit from some of the same things we are doing for TCP.

**Comment by quasiben** [comment]
add to allowlist

**Comment by gjoseph92** [comment]
@jakirkham thanks for the _pointer_ 😆 about NumPy! That is exactly the sort of thing I was looking for.

Using that I was able to come up with a much more sensible implementation of this in a new PR.

**Pull request closed without merging**
